Adds docs on migration from 3rd party networks to TTES (#3000)
TheThingsNetwork/lorawan-stack: doc/content/getting-started/migrating-from-networks/device-json.md

---
title: "Create Device.json"
description: ""
weight: 10
---

> {{% tts %}} allows you to import devices from other networks using a JSON file describing those devices. Devices imoported this way can be migrated without the need for a rejoin.

Reviewer: typo: imoported

## Required Fields in `devices.json`

> - [device_id]({{< ref "reference/glossary#device-id" >}}) 

Reviewer: Perhaps this list could be a table instead? cc @benolayinka

Author: Good idea.

- [application_id]({{< ref "reference/glossary#application-id" >}}) 
- [dev_eui]({{< ref "reference/glossary#deveui" >}}) 
- [join_eui]({{< ref "reference/glossary#joineui" >}}) (also referred to as AppEUI)
- name (optional)
- [lorawan_version]({{< ref "reference/glossary#lorawan-version" >}}) 
- [lorawan_phy_version]({{< ref "reference/glossary#regional-parameters" >}}) (or regional parameters)
- [frequency_plan_id]({{< ref "reference/glossary#frequency-plan" >}})
- supports_join (boolean `true` for OTAA, `false` for ABP devices)
- [app_key]({{< ref "reference/glossary#application-key" >}})
- nwk_key (only for LoRaWAN version 1.1+)

> - [rx1_delay]({{< ref "reference/api/end_device#message:MACSettings" >}}) (optional). Choose between `RX_DELAY_1` or `RX_DELAY_5` for an RX1 delay of 1 or 5 seconds

Reviewer: ```suggestion
- [rx1_delay]({{< ref "reference/api/end_device#message:MACSettings" >}}) (optional). Typical values are `RX_DELAY_1` (1 second) and `RX_DELAY_5` (5 seconds)
```

- [supports_32_bit_f_cnt]({{< ref "reference/api/end_device#message:MACSettings" >}}) (boolean `true` for 32 bit frame counters, `false` for non-32 bit counters)
- [dev_addr]({{< ref "reference/glossary#devaddr" >}})
- [app_s_key]({{< ref "reference/glossary#application-session-key" >}})
- [f_nwk_s_int_key]({{< ref "reference/api/end_device#message:SessionKeys" >}}) (also referred to as Network Session Key)
- [last_f_cnt_up]({{< ref "reference/api/end_device#message:MACSettings" >}}) (optional, frame counter uplinks)
- [last_n_f_cnt_down]({{< ref "reference/api/end_device#message:MACSettings" >}}) (optional, frame counter downlinks)

## Example `devices.json`

> Below is an example `devices.json` file. The file may contain multiple devices, stored as an array of the following JSON object.

Reviewer: > stored as an array of the following JSON object

This is misleading. The file should contain multiple JSON objects, not an array of them.

NOTE: A JSON array is

```
[
  {... device1 ...}, 
  {... device2 ...},
]
```

Whereas `ttn-lw-cli` expects:

```
{... device1 ...}
{... device2 ...}
```

@benolayinka Any suggestion about how to make this clearer in the documentation?

```json
{
  "ids": {

>     "device_id": "device_id",

Reviewer: `_` is an illegal character for IDs, so suggest something else instead.

```suggestion
    "device_id": "device-id",
```

    "application_ids": {

>       "application_id": "application_id"

Reviewer: same here

    },
    "dev_eui": "0000000000000000",
    "join_eui": "0000000000000000"
  },

>   "name": "name_of_device",

Reviewer: I would suggest including the `description` field here as well

  "lorawan_version":"1.0.2",
  "lorawan_phy_version":"1.0.2-b",
  "frequency_plan_id":"EU_863_870",
  "supports_join":true,
  "root_keys":{
      "app_key":{
          "key":"00000000000000000000000000000000"
        }
      },
      "mac_settings":{
          "rx1_delay":{
              "value":"RX_DELAY_1"
              },
            "supports_32_bit_f_cnt":true
        },
        "session":{
            "dev_addr":"00000000",
            "keys":{
                "app_s_key":{
                    "key":"00000000000000000000000000000000"
                },
                "f_nwk_s_int_key":{
                    "key":"00000000000000000000000000000000"
                }
            },
        "last_f_cnt_up":0,
        "last_n_f_cnt_down":0

>     }

Reviewer: Indent is a bit off here

Reviewer: Please fix indentation as well.

## Required Fields in `devices.json`

| Field | Type | Description |
|---|---|---|---|
| `device_id` | string | [More info]({{< ref "reference/glossary#device-id" >}}) |
| `application_id` | string | [More info]({{< ref "reference/glossary#application-id" >}}) |
| `dev_eui` | uint64 | [More info]({{< ref "reference/glossary#deveui" >}}) |
| `join_eui` | uint64 | Also referred to as **AppEUI**. [More info]({{< ref "reference/glossary#joineui" >}}) |
| `name` | string | Optional, name of the device |
| `description` | string | Optional, description of the device |
| `lorawan_version` | `defined_only` | e.g.  MAC_V1_0_2`. [More info]({{< ref "reference/glossary#lorawan-version" >}}) |
| `lorawan_phy_version` | `defined_only` | e.g.  PHY_V1_0_2_REV_B`. Also referred to as **Regional Parameters Version**. [More info]({{< ref "reference/glossary#regional-parameters" >}}) |
| `frequency_plan_id` | `defined_only` | e.g.  EU_863_870`. [More info]({{< ref "reference/glossary#frequency-plan" >}}) |
| `supports_join` | boolean | `true` for OTAA, `false` for ABP devices |
| `app_key` | uint128 | Application Key. [More info]({{< ref "reference/glossary#application-key" >}}) |

> | `nwk_key` | uint128 | Network Key. Only for LoRaWAN version 1.1+ |

Reviewer: `nwk_key` is not a top-level field in the JSON file, so this might become confusing for users. The example JSON below *should* help sort most of that out.

However, I would suggest that the field names here are changed to imply the "full" path. e.g. change `nwk_key` to `root_keys.nwk_key.key`, `device_id` to `ids.device_id` etc. Maybe also add a comment for the meaning of the dot notation.

Adding the comment under `nwk_key`, but this obviously applies to most of the keys here.

Author: @neoaggelos not sure if I understand your comment.

What do you mean with *top-level field*?

Why would someone rather use `root_keys.nwk_key.key` instead of `nwk_key`, or `ids.device_id` instead of `device_id`? 

And, what is the meaning of the dot notation?

Reviewer: The problem is that a user that sees `nwk_key` in the table above, they may be misled to believe that the JSON file should have the following format:

```js
{
  "ids": {
    "device_id": "dev1",
     /* .... */
   },
  "nwk_key": "<NWK_KEY_HERE>",
  /* .... */
}
```

However, that is wrong, and they should instead do:

```js
{
  "ids": {
    "device_id": "dev1",
    /* .... */
  },
  "root_keys": {
    "nwk_key:": {
      "key": "<NWK_KEY_HERE>"
    }
  },
  /* ... */
}
```

The "dot notation" I refer to specifies the full path to the key. Hope this makes things a bit more clear.

Author: Thanks, this clarifies.

| `mac_settings.rx1_delay.value` | `RxDelayValue` | Optional. Typical values are `RX_DELAY_1` (1 second) and `RX_DELAY_5` (5 seconds).  [More info]({{< ref "reference/api/end_device#message:MACSettings" >}})|
| `mac_settings.supports_32_bit_f_cnt` | boolean | `true` for 32 bit frame counters, `false` for non-32 bit counters). [More info]({{< ref "reference/api/end_device#message:MACSettings" >}})  |
| `session.dev_addr` | uint32 | Device Address. [More info]({{< ref "reference/glossary#devaddr" >}}) |
| `session.keys.app_s_key.key` | uint128 | Application Session Key. [More info]({{< ref "reference/glossary#application-session-key" >}}) |
| `session.keys.f_nwk_s_int_key.key` | uint128 | Forwarding Network Session Integrity Key, also referred to as **Network Session Key** in LoRaWAN v1.0.x compatibility mode. [More info]({{< ref "reference/api/end_device#message:SessionKeys" >}}) |
| `session.last_f_cnt_up` | int | Optional, frame counter uplink. [More info]({{< ref "reference/api/end_device#message:MACSettings" >}}) |
| `session.last_n_f_cnt_down` | int | Optional, frame counter downlinks. [More info]({{< ref "reference/api/end_device#message:MACSettings" >}}) |

> > Note: the column **Field** uses *Dot Notations*, the below examples uses *Bracket Notation*. `root_keys.nwk_app.key` can also be written as: 

Reviewer: ```suggestion
> Note: Dots in the **Field** names imply an embedded object. For example, `root_keys.nwk_app.key` must be set as: 
```
